src/Http/Server.php

```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http;
+
+use Cake\Event\EventDispatcherTrait;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use RuntimeException;
+use Zend\Diactoros\Response;
+use Zend\Diactoros\Response\EmitterInterface;
+use Zend\Diactoros\Response\SapiStreamEmitter;
+
+/**
+ * Runs an application invoking all the PSR7 middleware and the registered application.
+ */
+class Server
+{
+
+    use EventDispatcherTrait;
+
+    /**
+     * @var \Cake\Http\BaseApplication
+     */
+    protected $app;
+
+    /**
+     * @var \Cake\Http\Runner
+     */
+    protected $runner;
+
+    /**
+     * Constructor
+     *
+     * @param \Cake\Http\BaseApplication $app The application to use.
+     */
+    public function __construct(BaseApplication $app)
+    {
+        $this->setApp($app);
+        $this->setRunner(new Runner());
+    }
+
+    /**
+     * Run the request/response through the Application and its middleware.
+     *
+     * This will invoke the following methods:
+     *
+     * - App->bootstrap() - Perform any bootstrapping logic for your application here.
+     * - App->middleware() - Attach any application middleware here.
+     * - Trigger the 'Server.buildMiddleware' event. You can use this to modify the
+     *   from event listeners.
+     * - Run the middleware stack including the application.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request  The request to use or null.
+     * @param \Psr\Http\Message\ResponseInterface      $response The response to use or null.
+     * @return \Psr\Http\Message\ResponseInterface
+     * @throws \RuntimeException When the application does not make a response.
+     */
+    public function run(ServerRequestInterface $request = null, ResponseInterface $response = null)
+    {
+        $this->app->bootstrap();
+        $request = $request ?: ServerRequestFactory::fromGlobals();
+        $response = $response ?: new Response();
+
+        $middleware = $this->app->middleware(new MiddlewareStack());
+        if (!($middleware instanceof MiddlewareStack)) {
+            throw new RuntimeException('The application `middleware` method did not return a middleware stack.');
+        }
+        $middleware->push($this->app);
+        $this->dispatchEvent('Server.buildMiddleware', ['middleware' => $middleware]);
```

Comment:
I figured plugins might want a way to hook into the application middleware being built for zero config style plugins.